### PR TITLE
Input, Checkbox, Switch and Radio converted to TypeScript

### PR DESCRIPTION
**Pull request by alexasselin008**

### CODE REVIEW

File: packages/react-components/src/radio/src/RadioGroup.tsx [hunk 1]
```diff
@@ -120,6 +124,9 @@ export function InnerRadioGroup(props) {
         disabled,
         children,
         forwardedRef,
+        // we don't want size passed down to Group, since
```
Comment (alexasselin008) [inline packages/react-components/src/radio/src/RadioGroup.tsx:127]: size used to be in the rest property, but nobody down the line supports this props, we pass it down to a Box, whos passes it down to a div, but its not supported there.

File: packages/react-components/src/radio/src/RadioGroup.tsx [hunk 2]
```diff
@@ -158,12 +165,12 @@ export function InnerRadioGroup(props) {
     const { groupProps, itemProps } = useGroupInput({
         cssModule: "o-ui-radio-group",
         role: "radio-group",
-        keyProp: KeyProp,
-        value,
-        defaultValue,
+        // keyProp: KeyProp, //TODO: why is this passed but not used?
```
Comment (alexasselin008) [inline packages/react-components/src/radio/src/RadioGroup.tsx:168]: useGroupInput doesn't use those property, what do we do?

File: packages/react-components/src/radio/src/RadioGroup.tsx [hunk 3]
```diff
@@ -158,12 +165,12 @@ export function InnerRadioGroup(props) {
     const { groupProps, itemProps } = useGroupInput({
         cssModule: "o-ui-radio-group",
         role: "radiogroup",
-        keyProp: RadioKeyProp,
-        value,
-        defaultValue,
+        // keyProp: RadioKeyProp, //TODO: why is this passed but not used?
```
Comment (alexasselin008) [inline packages/react-components/src/radio/src/RadioGroup.tsx:168]: why are we forwarding useless props? GroupInput doesnt use them

File: packages/react-components/src/checkbox/src/Checkbox.tsx [hunk 4]
```diff
@@ -1,79 +1,90 @@
...
     /**
      * The initial value of `indeterminate`.
      */
-    defaultIndeterminate: bool,
+    defaultIndeterminate?: boolean;
     /**
      * The value to associate with when in a group.
      */
-    value: oneOfType([string, number]),
+    value?: string | number;
```
Comment (patricklafrance) [inline packages/react-components/src/checkbox/src/Checkbox.tsx:42]: I forgot why a checkbox accept a "number". 

[MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox) specify that a checkbox `value` prop is a DOMString so it doesn't make sense.

I think we should only accept a "string" as value.

Is this something you would to take care of?

File: packages/react-components/src/input/src/wrappedInputPropsAdapter.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        ...fieldProps,
+        wrapperProps: {
+            className
+        }
+    };
+
+    return [props, isInField];
+}
+
+export function wrappedInputPropsAdapter({ className, ...rest }: Record<string, any>) {
```
Comment (patricklafrance) [inline packages/react-components/src/input/src/wrappedInputPropsAdapter.ts:22]: I believe it should have the same return type as useFieldWrappedInput (minus the array of course)

File: packages/react-components/src/input/src/wrappedInputPropsAdapter.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { UseFieldInputPropsReturn, useFieldInputProps } from "../../field";
+
+export interface FieldWrappedInputProps extends Omit<UseFieldInputPropsReturn[0], "className"> {
```
Comment (patricklafrance) [inline packages/react-components/src/input/src/wrappedInputPropsAdapter.ts:3]: This array index thing is weird.

Looking at the code in ``FieldContext.tsx` I would cheat a little bit and do:

```
export type UseFieldInputPropsReturn = {
    validationState?: "valid" | "invalid";
    id?: string;
    required?: boolean;
    disabled?: boolean;
    fluid?: boolean;
    size?: "sm" | "md";
    className?: string;
    "aria-labelledby"?: string;
    "aria-describedby"?: string;
};

export function useFieldInputProps(): [UseFieldInputPropsReturn, boolean] {
```

File: packages/react-components/src/radio/src/Radio.tsx [hunk 7]
```diff
@@ -17,51 +17,68 @@ import {
...
     /**
      * The initial value of `checked` when uncontrolled.
      */
-    defaultChecked: bool,
+    defaultChecked?: boolean;
     /**
      * The value to associate with when in a group.
      */
-    value: oneOfType([string, number]).isRequired,
+    value: string | number
```
Comment (patricklafrance) [inline packages/react-components/src/radio/src/Radio.tsx:36]: Same as checkbox, as per [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/radio) the value should only be a string.

(Would also requires a few changes like Checkbox, let me know if you want to take care of it)

File: packages/react-components/src/radio/src/RadioGroup.tsx (hunk 3 above)
Comment (patricklafrance) [reply packages/react-components/src/radio/src/RadioGroup.tsx:168]: Bad refactor I guess, remove them. I think useGroupInput used to "spread" the rest.

File: packages/react-components/src/radio/src/RadioGroup.tsx [hunk 8]
```diff
@@ -133,7 +140,7 @@ export function InnerRadioGroup(props) {
 
     const groupRef = useMergedRefs(setFocusRef, forwardedRef);
 
-    const handleArrowSelect = useEventCallback((event, element) => {
+    const handleArrowSelect = useEventCallback((_event, element) => {
```
Comment (patricklafrance) [inline packages/react-components/src/radio/src/RadioGroup.tsx:143]: Why the "_" ?

File: packages/react-components/src/radio/src/RadioGroup.tsx (hunk 1 above)
Comment (patricklafrance) [reply packages/react-components/src/radio/src/RadioGroup.tsx:127]: Why size would be in the "rest" prop? Pretty sure a RadioGroup or Radio doesn't support `size`.

File: packages/react-components/src/shared/src/checkable.ts [hunk 9]
```diff
@@ -1,9 +1,11 @@
 import { ChangeEvent, createContext, useContext } from "react";
 import { isNil } from "lodash";
 
+type CheckableValueType = boolean | string | number | CheckableValueType[]
```
Comment (patricklafrance) [inline packages/react-components/src/shared/src/checkable.ts:4]: CheckableValueType > CheckableValue ?

Comment (patricklafrance) [reply packages/react-components/src/shared/src/checkable.ts:4]: Was looking at

```
type CheckableProps = Omit<CheckableContextType, "checkedValue"> & {
    checked?: boolean;
}

export function useCheckableProps({ value }: { value?: boolean | string | number }): [CheckableProps, boolean] {
```

To be consistant, CheckableProps  -> UseCheckablePropsReturn ? Not sure about our naming conventions here we seem to diverge.

File: packages/react-components/src/shared/src/useRovingFocus.ts [hunk 10]
```diff
@@ -57,7 +57,7 @@ export function useRovingFocus(scope: DomScope, { isDisabled = false } = {}) {
 /*
 IMPORTANT: Keyed roving focus doesn't handle disabled elements. This is the responsability of the calling component to ensure that the `currentKey` doesn't match a disabled element.
 */
-export function useKeyedRovingFocus(scope: DomScope, currentKey: string, { keyProp = "value", isDisabled = false } = {}) {
+export function useKeyedRovingFocus(scope: DomScope, currentKey: string | number, { keyProp = "value", isDisabled = false } = {}) {
```
Comment (patricklafrance) [inline packages/react-components/src/shared/src/useRovingFocus.ts:60]: I guess the "string | number" would change once Checkbox and Radio are fixed

File: packages/react-components/src/checkbox/src/Checkbox.tsx (hunk 4 above)
Comment (alexasselin008) [reply packages/react-components/src/checkbox/src/Checkbox.tsx:42]: i don't mind doing the change in this PR. However, according to the doc you linked, any value other than DOMString will be stringified, so it's not exactly true that it only accept a string. According to Typescript's typing of an input, the allowed values are : `value?: string | ReadonlyArray<string> | number;`

File: packages/react-components/src/input/src/wrappedInputPropsAdapter.ts (hunk 6 above)
Comment (alexasselin008) [reply packages/react-components/src/input/src/wrappedInputPropsAdapter.ts:3]: done

File: packages/react-components/src/input/src/wrappedInputPropsAdapter.ts (hunk 5 above)
Comment (alexasselin008) [reply packages/react-components/src/input/src/wrappedInputPropsAdapter.ts:22]: should
```

export function useFieldWrappedInput(): [FieldWrappedInputProps, boolean] {
    const [{ className, ...fieldProps }, isInField] = useFieldInputProps();

    const props = {
        ...fieldProps,
        wrapperProps: {
            className
        }
    };

    return [props, isInField];
}
```

become 
```

export function useFieldWrappedInput(): [FieldWrappedInputProps, boolean] {
    const [fieldProps, isInField] = useFieldInputProps();

    const props =wrappedInputPropsAdapter(fieldProps);

    return [props, isInField];
}
```

File: packages/react-components/src/radio/src/RadioGroup.tsx (hunk 1 above)
Comment (alexasselin008) [reply packages/react-components/src/radio/src/RadioGroup.tsx:127]: My fix was a bad one. 

the correct fix would be 
```
const {
        value,
        defaultValue,
        required,
        validationState,
        name,
        onChange,
        autoFocus,
        orientation = "vertical",
        gap,
        wrap,
        reverse,
        disabled,
        children,
        forwardedRef,
        ...rest
    } = mergeProps(
        props,
        toolbarProps,
        omitProps(fieldProps, ["fluid", "size"])  <---- fieldProps contains a size, but size is not supported by <Group
    );
```

Comment (patricklafrance) [reply packages/react-components/src/radio/src/RadioGroup.tsx:127]: Sounds good

File: packages/react-components/src/input/src/wrappedInputPropsAdapter.ts (hunk 5 above)
Comment (patricklafrance) [reply packages/react-components/src/input/src/wrappedInputPropsAdapter.ts:22]: Sounds good

File: packages/react-components/src/radio/src/RadioGroup.tsx (hunk 3 above)
Comment (alexasselin008) [reply packages/react-components/src/radio/src/RadioGroup.tsx:168]: done

File: packages/react-components/src/shared/src/checkable.ts (hunk 9 above)
Comment (alexasselin008) [reply packages/react-components/src/shared/src/checkable.ts:4]: done x2

File: packages/react-components/src/input/src/wrappedInputPropsAdapter.ts (hunk 6 above)
Comment (alexasselin008) [reply packages/react-components/src/input/src/wrappedInputPropsAdapter.ts:3]: done

File: packages/react-components/src/radio/src/Radio.tsx (hunk 7 above)
Comment (alexasselin008) [reply packages/react-components/src/radio/src/Radio.tsx:36]: in other pr

**Review by patricklafrance — APPROVED** [review]

**Pull request merged**